### ISSUE TITLE
C++11 flags not set with /bootstrap

### ISSUE THREAD

**Issue by bendikro** [arvidn/libtorrent#870]
**Please provide the following information**

libtorrent version (or branch):
master

platform/architecture:
Ubuntu 16.04, x64

compiler and compiler version:
gcc 5.3.1

After ./bootstrap, compiling fails with 
```
/usr/include/c++/5/bits/c++0x_warning.h:32:2: error: #error This file requires compiler and library support for the ISO C++ 2011 standard. This support must be enabled with the -std=c++11 or -std=gnu++11 compiler options.
```

Fixed by changing this in configure.ac:
```
-CXXFLAGS="$CXXFLAGS -ftemplate-depth=120"
+CXXFLAGS="$CXXFLAGS -ftemplate-depth=200 -std=c++11"
 ```
It was also necessary to increase the ftemplate-depth to at least 123 compile successfully. Not sure what value you'd prefer.

To compile python bindings successfully:
```
 dnl Pass some build options to .pc file
-COMPILETIME_OPTIONS=""
+COMPILETIME_OPTIONS="-std=c++11"
```

**Comment by arvidn** [comment]
does there need to be any conditionals around this to only apply it when using gcc or clang?
in other words, what happens when using icc or sunpro for instance.

**Comment by bendikro** [comment]
I've only tested with gcc (and now also clang which gives the same error) so I don't know tbh.

**Comment by arvidn** [comment]
are there any conventions how to encode compiler specific options in autoconf/automake files?

**Comment by bendikro** [comment]
I found a couple of helpers for this, but beware, I've only tested with gcc/clang.

* `AX_CXX_COMPILE_STDCXX(11, noext, mandatory)` (https://www.gnu.org/software/autoconf-archive/ax_cxx_compile_stdcxx.html)

 This will odly enough modify **CXX** and **CXXCPP** variables, but it works. However, the flag is not written to the `bindings/python/compiler_flags` which depends on **COMPILETIME_OPTIONS** and **CPPFLAGS**. Therefore some extra logic would be required to use this macro.

* `AX_CHECK_COMPILE_FLAG()` (https://www.gnu.org/software/autoconf-archive/ax_check_compile_flag.html#ax_check_compile_flag)
 Lets you check specific compiler flags.

A solution with `AX_CHECK_COMPILE_FLAG()` could be:
```diff
CC="$PTHREAD_CC"
-CXXFLAGS="$CXXFLAGS -ftemplate-depth=120"
+CXXFLAGS="$CXXFLAGS -ftemplate-depth=200"
+AX_CHECK_COMPILE_FLAG([-std=c++11], [CPPFLAGS="$CPPFLAGS -std=c++11"])
```
Doing this will work, however it might be a bit hacky to use the **CPPFLAGS** variable meant for the C PreProcessor?

A solution that doesn't use **CPPFLAGS** would be something like this:
```diff
 CC="$PTHREAD_CC"
-CXXFLAGS="$CXXFLAGS -ftemplate-depth=120"
+CXX_STD11_FLAG=""
+AX_CHECK_COMPILE_FLAG([-std=c++11], [CXX_STD11_FLAG="$CXX_STD11_FLAG -std=c++11"])
+CXXFLAGS="$CXXFLAGS $CXX_STD11_FLAG -ftemplate-depth=200"
```
 ....
```diff
dnl Pass some build options to .pc file
-COMPILETIME_OPTIONS=""
+COMPILETIME_OPTIONS="$CXX_STD11_FLAG"
```

**Comment by tete1030** [comment]
On OS X El Capitan, in addition to `-std=c++11`, I also need to add `-stdlib=libc++ -mmacosx-version-min=10.7` to the `COMPILETIME_OPTIONS`, or it will report
```
In file included from /usr/local/opt/boost/include/boost/bind.hpp:22:
/usr/local/opt/boost/include/boost/bind/bind.hpp:880:53: error: no template
      named 'forward' in namespace 'std'; did you mean simply 'forward'?
```
when compiling python bindings.

`-mmacosx-version-min=10.7` is necessary or 
```
clang: error: invalid deployment target for -stdlib=libc++ (requires OS X 10.7 or later)
```

**Comment by arvidn** [comment]
I have added ax_cxx_compile_stdcxx_11.m4 and detection for C++11 support to master. For some reason the -std=c++11 flag is added directly to the CXX variable though, as @bendikro points out. I worked around this by setting the CC variable to the CXX variable in setup.py. It seems python's distutils doesn't use CXX. It appears to be working on linux now at least as the autotools now are tested on travis
